src/Data/Parameterized/NatRepr.hs

```diff
@@ -176,6 +177,14 @@ knownNat = go Proxy
         go p = assert (v <= maxInt) (NatRepr (fromInteger v))
           where v = natVal p
 
+withKnownNat :: forall n r. NatRepr n -> (KnownNat n => r) -> r
+withKnownNat (NatRepr nVal) v =
+  case someNatVal nVal of
+    Just (SomeNat (_ :: Proxy n')) ->
```

Comment:
It probably doesn't make any real difference, but I would feel slightly better if the pattern match also forced the `Proxy` value, just to reinforce that there are no games being played with `undefined`.